exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java

```diff
@@ -130,562 +145,248 @@ public IterOutcome innerNext() {
-          }
-        } else {
-          IterOutcome iterOutcome = leftSide.nextBatch();
-
-          switch(iterOutcome) {
-            case STOP:
-            case OUT_OF_MEMORY:
-              upstream = iterOutcome;
-              return upstream;
-
-            case OK_NEW_SCHEMA:
-              if(!leftSide.getRecordBatch().getSchema().equals(leftSchema)) {
-                throw new SchemaChangeException("Schema change detected in the left input of Union-All. This is not currently supported");
-              }
-
-              iterOutcome = IterOutcome.OK;
-              // fall through
-            case OK:
-              unionAllRecordBatch.setCurrentRecordBatch(leftSide.getRecordBatch());
-              upstream = iterOutcome;
-              return upstream;
-
-            case NONE:
-              unionAllRecordBatch.setCurrentRecordBatch(rightSide.getRecordBatch());
-              upstream = IterOutcome.OK;
-              leftIsFinish = true;
-              return upstream;
-
-            default:
-              throw new IllegalStateException(String.format("Unknown state %s.", upstream));
-          }
-        }
-      }
-    }
 
-    /**
-     *
-     * Summarize the inference in the four different situations:
-     * First of all, the field names are always determined by the left side
-     * (Even when the left side is from an empty file, we have the column names.)
-     *
-     * Cases:
-     * 1. Left: non-empty; Right: non-empty
-     *      types determined by both sides with implicit casting involved
-     * 2. Left: empty; Right: non-empty
-     *      type from the right
-     * 3. Left: non-empty; Right: empty
-     *      types from the left
-     * 4. Left: empty; Right: empty
-     *      types are nullable integer
-     */
-    private void inferOutputFields() {
-      if(!leftIsFinish && !rightIsFinish) {
-        // Both sides are non-empty
-        inferOutputFieldsBothSide();
-      } else if(!rightIsFinish) {
-        // Left side is non-empty
-        // While use left side's column names as output column names,
-        // use right side's column types as output column types.
-        inferOutputFieldsFromSingleSide(
-            leftSide.getRecordBatch().getSchema(),
-            rightSide.getRecordBatch().getSchema());
+  // The output table's column names always follow the left table,
+  // where the output type is chosen based on DRILL's implicit casting rules
+  private void inferOutputFieldsBothSide(final BatchSchema leftSchema, final BatchSchema rightSchema) {
+//    outputFields = Lists.newArrayList();
+    final Iterator<MaterializedField> leftIter = leftSchema.iterator();
+    final Iterator<MaterializedField> rightIter = rightSchema.iterator();
+
+    int index = 1;
+    while (leftIter.hasNext() && rightIter.hasNext()) {
+      MaterializedField leftField  = leftIter.next();
+      MaterializedField rightField = rightIter.next();
+
+      if (hasSameTypeAndMode(leftField, rightField)) {
```

Comment:
The code here has the same "flavor" as the code in the code gen setup. Maybe add an explanation for why we need both flavors? What does the one do that the other does not do?

Reply:
If you are referring `createUnionAller` and `inferOutputFieldsBothSide`,  the majority of code actually are identical from the original one. AFASIK, the later one is to determine the output field name/type, while the first one is to generate code (may require implicit cast, depending on the future incoming batch's schema and output field type).  

1. https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java#L243-L244
2. https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java#L582-L585